File: src/main/resources/io/viash/languages/r/ViashParseJson.R
#' Parse JSON parameters file into an R list using jsonlite
#'
#' Uses the jsonlite package for JSON parsing. Requires jsonlite to be installed.
#'
#' @param json_path Path to the JSON file. If NULL, reads from
#'   $VIASH_WORK_PARAMS environment variable.
#' @return Named list containing the parsed JSON data.
viash_parse_json <- function(json_path = NULL) {
  if (is.null(json_path)) {
    json_path <- Sys.getenv("VIASH_WORK_PARAMS")
    if (json_path == "") {
      stop("VIASH_WORK_PARAMS environment variable not set")
    }
  }

  if (!file.exists(json_path)) {
    stop(paste0("Parameters file not found: ", json_path))
  }

  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    stop(
      "The 'jsonlite' R package is required but not installed. ",
      "Please install it with: install.packages('jsonlite')"
    )
  }

  tryCatch({
    json_text <- readLines(json_path, warn = FALSE)
    json_text <- paste(json_text, collapse = "\n")
    # Use bigint_as_char = TRUE to preserve large integers as strings
    # This prevents precision loss for values > 2^53
    jsonlite::parse_json(json_text, simplifyVector = TRUE, bigint_as_char = TRUE)
  }, error = function(e) {
    stop(paste0("Error parsing JSON file: ", e$message))
  })
}

File: src/main/resources/io/viash/languages/javascript/ViashParseJson.js
const fs = require('fs');

/**
 * Parse JSON parameters file into a JavaScript object.
 * 
 * @param {string|null} jsonPath - Path to the JSON file. If null, reads from $VIASH_WORK_PARAMS environment variable.
 * @returns {Object} Parsed JSON data.
 */
function viashParseJson(jsonPath = null) {
  if (jsonPath === null) {
    jsonPath = process.env.VIASH_WORK_PARAMS;
    if (!jsonPath) {
      throw new Error("VIASH_WORK_PARAMS environment variable not set");
    }
  }
  
  if (!fs.existsSync(jsonPath)) {
    throw new Error(`Parameters file not found: ${jsonPath}`);
  }
  
  try {
    const jsonText = fs.readFileSync(jsonPath, 'utf8');
    return JSON.parse(jsonText);
  } catch (error) {
    throw new Error(`Error parsing JSON file: ${error.message}`);
  }
}

module.exports = { viashParseJson };
